samples: Add red-black tree sample program

rruuaanng opened the pull request:
Introduce an example program to demonstrate
how to use the red-black tree API.

kartben reviewed (changes requested):
should be in `samples/data_structures` and have a proper README
Also would suggest adding code comments to the various "sections" of your sample (and/or use printk's), otherwise it's not very educational IMO
It also needs a sample.yaml so that it's built in CI (and you can have it actually run and be tested on native_sim)

rruuaanng committed:
samples: data_structures: Fix the incorrect sample path in README

Change `samples/lib/mini_heap` to the actual path
`samples/data_structures/mini_heap`.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
samples: Add red-black tree sample program

Introduce an example program to demonstrate
how to use the red-black tree API.

Signed-off-by: James Roy <rruuaanng@outlook.com>
